src/main/java/picard/illumina/IlluminaBasecallsToFastq.java

```diff
@@ -504,6 +531,116 @@ private void makeFastqRecords(final FastqRecord[] recs, final int[] indices,
+        private final int numReads;
+
+        public ClusterWriter(final File[] templateFiles,
+                             final File[] sampleBarcodeFiles,
+                             final File[] molecularBarcodeFiles) {
+
+            this.templateOut = Arrays.stream(templateFiles).map(this::makeWriter).toArray(OutputStream[]::new);
+            this.sampleBarcodeOut = Arrays.stream(sampleBarcodeFiles).map(this::makeWriter).toArray(OutputStream[]::new);
+            this.molecularBarcodeOut = Arrays.stream(molecularBarcodeFiles).map(this::makeWriter).toArray(OutputStream[]::new);
+            this.appendTemplateNumber = this.templateOut.length > 1;
+            this.appendMolecularBarcodeNumber = this.molecularBarcodeOut.length > 1;
+            this.numReads = templateOut.length + sampleBarcodeOut.length + molecularBarcodeOut.length;
+        }
+
+        private OutputStream makeWriter(final File file) {
+            OutputStream os = Defaults.CREATE_MD5 ? IOUtil.openFileForMd5CalculatingWriting(file) : IOUtil.openFileForWriting(file);
+            os = IOUtil.maybeBufferOutputStream(os);
+            return os;
+        }
+
+        @Override
+        public void write(final ClusterData rec) {
+            int templateIndex = 0;
+            int sampleBarcodeIndex = 0;
+            int molecularBarcodeIndex = 0;
+
+            for (int i=0; i<this.numReads; ++i) {
+                final ReadData read = rec.getRead(i);
+                final OutputStream out;
+                final String name;
+
+                switch (read.getReadType()) {
+                    case T:
+                        out = templateOut[templateIndex++];
+                        name = readNameEncoder.generateReadName(rec, appendTemplateNumber ? templateIndex : null);
+                        break;
+                    case B:
+                        out = sampleBarcodeOut[sampleBarcodeIndex++];
+                        name = readNameEncoder.generateReadName(rec, null);
+                        break;
+                    case M:
+                        out = molecularBarcodeOut[molecularBarcodeIndex++];
+                        name = readNameEncoder.generateReadName(rec, appendMolecularBarcodeNumber ? molecularBarcodeIndex : null);
+                        break;
+                    default:
+                        throw new IllegalStateException("Read type other than T/B/M encountered.");
+                }
+
+                writeSingle(out, name, read);
+            }
+        }
+
+        /** Writes out a single read to a single FASTQ's output stream. */
+        private void writeSingle(final OutputStream out, final String name, final ReadData read) {
+            try {
+                final byte[] bases = read.getBases();
+                final byte[] quals = read.getQualities();
+                final int len = bases.length;
+                for (int i=0; i<len; ++i) {
+                    if (bases[i] == PERIOD) bases[i] = N;
```

Comment:
If we pushed down the "no call byte" to the BclReader we could avoid this recoding and just initially set no calls to `N` on the initial decode. 

Not sure it is entirely worth it as this is likely pretty fast.

Reply:
Agreed - and since we have to loop here to convert quals to fastq/+33 I don't think it would help all that much.